src/backend/distributed/metadata/distobject.c

```diff
@@ -0,0 +1,282 @@
+/*-------------------------------------------------------------------------
+ *
+ * distobject.c
+ *	  Functions to interact with distributed objects by their ObjectAddress
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "miscadmin.h"
+
+#include "access/genam.h"
+#include "access/heapam.h"
+#include "access/htup_details.h"
+#include "access/skey.h"
+#include "access/xact.h"
+#include "catalog/dependency.h"
+#include "catalog/namespace.h"
+#include "catalog/objectaddress.h"
+#include "catalog/pg_namespace.h"
+#include "catalog/pg_type.h"
+#include "distributed/metadata/distobject.h"
+#include "distributed/metadata/pg_dist_object.h"
+#include "distributed/metadata_cache.h"
+#include "executor/spi.h"
+#include "nodes/makefuncs.h"
+#include "nodes/pg_list.h"
+#include "parser/parse_type.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/regproc.h"
+#include "utils/rel.h"
+
+
+static int ExecuteCommandAsSuperuser(char *query, int paramCount, Oid *paramTypes,
+									 Datum *paramValues);
+
+
+PG_FUNCTION_INFO_V1(master_unmark_object_distributed);
+
+
+/*
+ * master_unmark_object_distributed(classid oid, objid oid, objsubid int)
+ *
+ * removes the entry for an object address from pg_dist_object. Only removes the entry if
+ * the object does not exist anymore.
+ */
+Datum
+master_unmark_object_distributed(PG_FUNCTION_ARGS)
+{
+	Oid classid = PG_GETARG_OID(0);
+	Oid objid = PG_GETARG_OID(1);
+	int32 objsubid = PG_GETARG_INT32(2);
+
+	ObjectAddress address = { 0 };
+	ObjectAddressSubSet(address, classid, objid, objsubid);
+
+	if (!IsObjectDistributed(&address))
+	{
+		/* if the object is not distributed there is no need to unmark it */
+		PG_RETURN_VOID();
+	}
+
+	if (ObjectExsists(&address))
```

Comment:
for the forceful unmark we could instruct them to use a super user account (unless they are in cloud) to remove the entries from `citus.pg_dist_object`.

I guess having a feature flag to prevent object creation on node activation could help in such cases, but then we get back into a situation where reference tables are causing issues.

Given this patch only creates schema's as a dependency (which was already the case before) I think it needs to be more granular then completely disabling object creation. Good thing to add to the types PR, a way to not follow certain edges of the dependency tree when the feature flag is turned off.